app/src/test/kotlin/com/exactpro/th2/lwdataprovider/grpc/GrpcDataProviderBackPressureTest.kt

```diff
@@ -0,0 +1,194 @@
+            endTimestamp.plusNanos(1),
+            endTimestamp.plus(5, ChronoUnit.MINUTES),
+        )
+        val group = "test"
+
+        whenever(storage.getGroupedMessageBatches(argThat {
+            groupName == group && from.value == startTimestamp && to.value == endTimestamp
+        })).thenReturn(ImmutableListCradleResult(firstBatches))
+
+        whenever(storage.getGroupedMessageBatches(argThat {
+            groupName == group && from.value == firstBatches.maxOf { it.lastTimestamp } && to.value == endTimestamp
+        })).thenReturn(ImmutableListCradleResult(lastBatches))
+
+        if(newDataOnEveryRequest) {
+            whenever(storage.getGroupedMessageBatches(argThat {
+                groupName == group  && to.value == endTimestamp
+            })).thenReturn(ImmutableListCradleResult(lastBatches))
+
+            whenever(storage.getGroupedMessageBatches(argThat {
+                limit == 1 && groupName == group
+            })).thenReturn(ImmutableListCradleResult(emptyBatch))
+        } else {
+            whenever(storage.getGroupedMessageBatches(argThat {
+                groupName == group  && to.value == endTimestamp
+            })).thenReturn(ImmutableListCradleResult(emptyBatch))
+
+            whenever(storage.getGroupedMessageBatches(argThat {
+                limit == 1 && groupName == group
+            })).thenReturn(ImmutableListCradleResult(emptyBatch))
+        }
+
+        val request = MessageGroupsSearchRequest.newBuilder().apply {
+            addMessageGroupBuilder().setName("test")
+            addResponseFormats(ResponseFormat.BASE_64.name)
+            bookIdBuilder.setName("test")
+            this.startTimestamp = startTimestamp.toTimestamp()
+            this.endTimestamp = endTimestamp.toTimestamp()
+            this.keepOpen = true
+        }.build()
+
+        val searchHandler = spy(searchHandler)
+        val requestContextCaptor = argumentCaptor<MessageResponseHandler>()
+        val grpcDataProvider = GrpcDataProviderBackPressure(
+            configuration,
+            searchHandler,
+            searchEventsHandler,
+            generalCradleHandler,
+            measurement,
+            backPressureExecutor
+        )
+        GrpcTestHolder(grpcDataProvider).use { (stub) ->
+
+            assertThrows<StatusRuntimeException> {
+                withDeadline(100) {
+                    val iterator = stub.searchMessageGroups(request)
+                    while (iterator.hasNext()) {
+                        iterator.next() // to not trigger backpressure
+                    }
+                }
+            }
+
+            verify(searchHandler).loadMessageGroups(
+                any(),
+                requestContextCaptor.capture(),
+                any()
+            )
+
+            val capturedRequestContext = requestContextCaptor.firstValue
+
+            assertInstanceOf<CancelableResponseHandler>(capturedRequestContext)
+
+            Thread.sleep(200)

```

Comment:
suggestion: you could use [awaitility](http://www.awaitility.org/) library instead of sleep call